reading_comprehension/qanet_encoder.py

```diff
@@ -0,0 +1,345 @@
+        self.self_attn_layer = MemoryEfficientMultiHeadSelfAttention(num_heads=num_attention_heads,
+                                                                     input_dim=hidden_dim,
+                                                                     attention_dim=attention_projection_dim,
+                                                                     values_dim=attention_projection_dim,
+                                                                     attention_dropout_prob=attention_dropout_prob)
+        self.ffn_norm_layer = LayerNorm(hidden_dim)
+        self.ffn = FeedForward(hidden_dim,
+                               activations=[Activation.by_name('relu')(),
+                                            Activation.by_name('linear')()],
+                               hidden_dims=[feedforward_hidden_dim, hidden_dim],
+                               num_layers=2,
+                               dropout=dropout_prob)
+
+        self.dropout = Dropout(dropout_prob)
+        self.residual_with_layer_dropout = ResidualWithLayerDropout(layer_dropout_undecayed_prob)
+        self._input_dim = input_dim
+        self._output_dim = hidden_dim
+
+    @overrides
+    def get_input_dim(self) -> int:
+        return self._input_dim
+
+    @overrides
+    def get_output_dim(self) -> int:
+        return self._output_dim
+
+    @overrides
+    def is_bidirectional(self):
+        return False
+
+    @overrides
+    def forward(self, inputs: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:  # pylint: disable=arguments-differ
+
+        if self._use_positional_encoding:
+            output = add_positional_features(inputs)
+        else:
+            output = inputs
+
+        total_sublayers = len(self._conv_layers) + 2
+        sublayer_cnt = 0
+
+        for conv_norm_layer, conv_layer in zip(self._conv_norm_layers, self._conv_layers):
+            conv_norm_out = self.dropout(conv_norm_layer(output))
+            conv_out = self.dropout(conv_layer(conv_norm_out.transpose_(1, 2)).transpose_(1, 2))
+            sublayer_cnt += 1
+            output = self.residual_with_layer_dropout(output, conv_out,
+                                                      sublayer_cnt, total_sublayers)
+
+        attn_norm_out = self.dropout(self.attn_norm_layer(output))
+        attn_out = self.dropout(self.self_attn_layer(attn_norm_out, mask))
+        sublayer_cnt += 1
+        output = self.residual_with_layer_dropout(output, attn_out,
+                                                  sublayer_cnt, total_sublayers)
+
+        ffn_norm_out = self.dropout(self.ffn_norm_layer(output))
+        ffn_out = self.dropout(self.ffn(ffn_norm_out))
+        sublayer_cnt += 1
+        output = self.residual_with_layer_dropout(output, ffn_out,
+                                                  sublayer_cnt, total_sublayers)
+
+        return output
+
+
+@Seq2SeqEncoder.register("stacked_qanet_encoder")
+class StackedQANetEncoder(Seq2SeqEncoder):
```

Comment:
`StackedQaNetEncoder`

Reply:
Actually, probably just `QaNetEncoder` - no need to have the `Stacked` in the name, as the `Block` in the other name already implies that.